### PR TITLE
ARROW-14759: [Doc] Steps in making your first PR - test in R

### PR DESCRIPTION
**Pull request by dragosmg**

### CODE REVIEW

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 1]
```diff
@@ -62,29 +62,81 @@ In this section we outline steps needed for unit testing in Arrow.
...
+      .. TODO make sure the format the dir tree below is formatted nicely (maybe as a code block)
+      tests
+        ├── testthat      # unit test files live here
+        └── testthat.R    # runs tests when R CMD check runs (e.g. with `devtools::check()`)
+
+      Usually, most files in the ``R/`` sub-folder have a corresponding test file in ``tests/testthat``.
+
+      **Running tests**
+
+      To run all tests in a package locally use ``devtools::test()`` in the R console. Alternatively, you can use ``make test`` in the shell.
```
Comment (dragosmg) [inline docs/source/developers/guide/step_by_step/testing.rst:104]: @jonkeane should we add more here about `make test`?

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 2]
```diff
@@ -62,29 +62,81 @@ In this section we outline steps needed for unit testing in Arrow.
...
+
+      ** Testing helpers**
+
+      To complement the ``testthat`` functionality, the ``arrow`` R package has defined a series of specific utility functions (called helpers), such as:
+
+      * Expectations - these start with ``expect_`` and are used to compare objects
+            - for example, ``expect_altrep_roundtrip()`` compares the result
+            of a function ``fn`` run on a vector ``x`` with the result of the
+            same function run on the altrep version of ``x``. More generally,
+            expect_…_roundtrip() functions do … .. TODO _fill int the blanks_
```
Comment (dragosmg) [inline docs/source/developers/guide/step_by_step/testing.rst:128]: @jonkeane I'm looking for a sentence here to summarise what the `expect_.._roundtrip()` family of functions does. Do you have any suggestions?

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 3]
```diff
@@ -62,29 +62,81 @@ In this section we outline steps needed for unit testing in Arrow.
...
-      .. TODO
+      We use `testthat <https://testthat.r-lib.org/index.html>`_ for unit testing in R. More specifically, we use the `3rd edition of testthat <https://testthat.r-lib.org/articles/third-edition.html>`_. On rare occasions we might want the behaviour of the 2nd edition of testthat, which is indicated by ``testthat::local_edition(2)``.
+
+      **Structure**
+
+      Expect the usual testthat folder structure:
+      .. TODO make sure the format the dir tree below is formatted nicely (maybe as a code block)
+      tests
+        ├── testthat      # unit test files live here
+        └── testthat.R    # runs tests when R CMD check runs (e.g. with `devtools::check()`)
```
Comment (AlenkaF) [inline docs/source/developers/guide/step_by_step/testing.rst:98]:
```suggestion
      
      .. code-block:: R
      
         tests
           ├── testthat      # unit test files live here
           └── testthat.R    # runs tests when R CMD check runs (e.g. with `devtools::check()`)
```

This works well in the html version of the docs.

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 4]
```diff
@@ -62,29 +62,81 @@ In this section we outline steps needed for unit testing in Arrow.
...
+
+      In general any change to source code needs to be accompanied by unit tests. All tests are expected to pass before a pull request is merged.
+
+      * Add functionality -> add unit tests
+      * Modify functionality -> update unit tests
+      * Solve a bug -> add unit test before solving it, which helps prove the bug and its fix
+      * Performance improvements should be reflected in benchmarks (which are also tests)
+      * An exception could be refactoring functionality that is fully covered by unit tests
+
+      ** Testing helpers**
```
Comment (AlenkaF) [inline docs/source/developers/guide/step_by_step/testing.rst:120]:
```suggestion
      **Testing helpers**
```

Comment (dragosmg) [reply docs/source/developers/guide/step_by_step/testing.rst]: Done

File: docs/source/developers/guide/step_by_step/testing.rst (hunk 3 above)
Comment (dragosmg) [reply docs/source/developers/guide/step_by_step/testing.rst]: Cool. Many thanks. Done

File: docs/source/developers/guide/step_by_step/testing.rst (hunk 1 above)
Comment (AlenkaF) [inline docs/source/developers/guide/step_by_step/testing.rst:105]:
```suggestion
      To run all tests in a package locally use this command in the R console:

      .. code-block:: R

         devtools::test()

      Alternatively, you can use this in the shell:

      .. code:: console

         $ make test
```

It is easier to find the necessary info in the html like so, with code being highlighted.

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 5]
```diff
@@ -62,29 +62,83 @@ In this section we outline steps needed for unit testing in Arrow.
...
+      We use `testthat <https://testthat.r-lib.org/index.html>`_ for unit testing in R. More specifically, we use the `3rd edition of testthat <https://testthat.r-lib.org/articles/third-edition.html>`_. On rare occasions we might want the behaviour of the 2nd edition of testthat, which is indicated by ``testthat::local_edition(2)``.
+
+      **Structure**
+
+      Expect the usual testthat folder structure:
+
+      .. code-block:: R
+      tests
+        ├── testthat      # unit test files live here
+        └── testthat.R    # runs tests when R CMD check runs (e.g. with `devtools::check()`)
```
Comment (AlenkaF) [inline docs/source/developers/guide/step_by_step/testing.rst:99]:
```suggestion
      .. code-block:: R

         tests
           ├── testthat      # unit test files live here
           └── testthat.R    # runs tests when R CMD check runs (e.g. with `devtools::check()`)
```

Needs an empty line and an indentation ;)

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 6]
```diff
@@ -62,29 +62,83 @@ In this section we outline steps needed for unit testing in Arrow.
...
+
+      To complement the ``testthat`` functionality, the ``arrow`` R package has defined a series of specific utility functions (called helpers), such as:
+
+      * Expectations - these start with ``expect_`` and are used to compare objects
+            - for example, ``expect_altrep_roundtrip()`` compares the result
+            of a function ``fn`` run on a vector ``x`` with the result of the
+            same function run on the altrep version of ``x``. More generally,
+            expect_…_roundtrip() functions do … .. TODO _fill int the blanks_
+            .. TODO
+            - Expect
```
Comment (AlenkaF) [inline docs/source/developers/guide/step_by_step/testing.rst:131]:
```suggestion

        - for example, ``expect_altrep_roundtrip()`` compares the result
          of a function ``fn`` run on a vector ``x`` with the result of the
          same function run on the altrep version of ``x``. More generally,
          ``expect_…_roundtrip()`` functions do …

        .. TODO _fill int the blanks_
        .. TODO
        
        - Expect
```

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 7]
```diff
@@ -62,29 +62,103 @@ In this section we outline steps needed for unit testing in Arrow.
...
+         devtools::test_active_file()
+
+      All tests are also run as part of our continuous integration (CI) pipelines.
+
+      **Good practice**
+
+      In general any change to source code needs to be accompanied by unit tests. All tests are expected to pass before a pull request is merged.
+
+      * Add functionality -> add unit tests
+      * Modify functionality -> update unit tests
```
Comment (jonkeane) [inline docs/source/developers/guide/step_by_step/testing.rst:133]: I like the simplicity of these bullet points, so maybe this should be in a footnote or in a paragraph below this, but we should also include something like:

If the new functionality is a user-facing or API change, you will almost certainly need to change tests — if no tests need to be changed it might mean the tests aren't right! If the new functionality is a refactor and no APIs are changing, there might not need to be test changes.

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 8]
```diff
@@ -62,29 +62,103 @@ In this section we outline steps needed for unit testing in Arrow.
...
+      * Solve a bug -> add unit test before solving it, which helps prove the bug and its fix
```
Comment (jonkeane) [inline docs/source/developers/guide/step_by_step/testing.rst:134]: 💯

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 9]
```diff
@@ -62,29 +62,103 @@ In this section we outline steps needed for unit testing in Arrow.
...
+      To complement the ``testthat`` functionality, the ``arrow`` R package has defined a series of specific utility functions (called helpers), such as:
+
+      * Expectations - these start with ``expect_`` and are used to compare objects
+            - for example, ``expect_altrep_roundtrip()`` compares the result
+            of a function ``fn`` run on a vector ``x`` with the result of the
+            same function run on the altrep version of ``x``. More generally,
+            expect_…_roundtrip() functions do …
+
+            .. TODO _fill in the blanks_
+            .. TODO
```
Comment (jonkeane) [inline docs/source/developers/guide/step_by_step/testing.rst:149]: I presume these will be done in another commit? OR should we remove them / add Jiras for following up?

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 10]
```diff
@@ -62,29 +62,103 @@ In this section we outline steps needed for unit testing in Arrow.
...
+        - ``skip_if_not_available()`` - another specific {arrow} skip. Arrow (libarrow) has a series of additional features that can be switched on or off (but this needs to happen at build time). If a unit test depends on such a feature and this feature is not available (i.e. was not selected when libarrow was built) the test is skipped, as opposed to having a failed test.
+        - ``skip_if_offline()`` - will not run tests that require an internet connection
+        - ``skip_on_os()`` - for unit tests that are OS specific.
+
+      *Important*: Once the conditions for a ``skip_()`` statement is met, no other line of code in the same ``test_that()`` test block will get executed.
```
Comment (jonkeane) [inline docs/source/developers/guide/step_by_step/testing.rst:159]:
```suggestion
      *Important*: Once the conditions for a ``skip_()`` statement is met, no other line of code in the same ``test_that()`` test block will get executed. If the skip is outside of a `test_that()` code block, it will skip the rest of the file.
```

**Review by jonkeane — CHANGES_REQUESTED** [review]
This is looking good, thank you! I've got a few comments / questions / additions

File: docs/source/developers/guide/step_by_step/testing.rst (hunk 1 above)
Comment (jonkeane) [reply docs/source/developers/guide/step_by_step/testing.rst]: A link to https://arrow.apache.org/docs/r/articles/developing.html#common-developer-workflow-tasks (or copying that content) might be nice.

File: docs/source/developers/guide/step_by_step/testing.rst (hunk 2 above)
Comment (jonkeane) [reply docs/source/developers/guide/step_by_step/testing.rst]: Ah I didn't see these questions in my first run though (thanks Github!)

It's kind of defining what roundtrip means, but I would say something like "They take an input, convert it to some other format (e.g. arrow) and then convert it back, confirming that the values are the same."

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 11]
```diff
@@ -62,29 +62,103 @@ In this section we outline steps needed for unit testing in Arrow.
...
+      We use `testthat <https://testthat.r-lib.org/index.html>`_ for unit testing in R. More specifically, we use the `3rd edition of testthat <https://testthat.r-lib.org/articles/third-edition.html>`_. On rare occasions we might want the behaviour of the 2nd edition of testthat, which is indicated by ``testthat::local_edition(2)``.
+
+      **Structure**
+
+      Expect the usual testthat folder structure:
+
+      .. code-block:: R
+
+         tests
+          ├── testthat      # unit test files live here
```
[the remark this answers is not in the document]
Comment (dragosmg) [reply docs/source/developers/guide/step_by_step/testing.rst:99]: Done

File: docs/source/developers/guide/step_by_step/testing.rst (hunk 2 above)
Comment (dragosmg) [reply docs/source/developers/guide/step_by_step/testing.rst]: Cool. Thanks.

File: docs/source/developers/guide/step_by_step/testing.rst (hunk 9 above)
Comment (dragosmg) [reply docs/source/developers/guide/step_by_step/testing.rst:149]: I will do them in another commit (incorporating your comments & suggestions)

File: docs/source/developers/guide/step_by_step/testing.rst (hunk 7 above)
Comment (dragosmg) [reply docs/source/developers/guide/step_by_step/testing.rst:133]: My last bullet point covers refactoring, but I've included your entire paragraph below the list. Done

File: docs/source/developers/guide/step_by_step/testing.rst (hunk 1 above)
Comment (dragosmg) [reply docs/source/developers/guide/step_by_step/testing.rst]: I linked to it since the information is slightly more in-depth and we do not need to keep the 2 testing docs in sync.

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 12]
```diff
@@ -62,29 +62,103 @@ In this section we outline steps needed for unit testing in Arrow.
...
+          └── testthat.R    # runs tests when R CMD check runs (e.g. with devtools::check())
```
[the remark this answers is not in the document]
Comment (dragosmg) [reply docs/source/developers/guide/step_by_step/testing.rst:100]: Done.

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 13]
```diff
@@ -62,29 +62,101 @@ In this section we outline steps needed for unit testing in Arrow.
...
+
+      You can run the tests in a single test file you have open with
+
+      .. code-block:: R
+
+         devtools::test_active_file()
+
+      All tests are also run as part of our continuous integration (CI) pipelines.
+
+      The Arrow R Developer guide also has a section on running tests. You can check it out `here <https://arrow.apache.org/docs/r/articles/developing.html#running-tests>`_.
```
Comment (jonkeane) [inline docs/source/developers/guide/step_by_step/testing.rst:130]:
```suggestion
      The `Arrow R Developer guide also has a section<https://arrow.apache.org/docs/r/articles/developing.html#running-tests>`_ on running tests.
```

Avoids the click here type of pattern (and is a bit more concise.

Comment (jonkeane) [reply docs/source/developers/guide/step_by_step/testing.rst:130]: I think I got the RST correct here, but not sure about that `_` at the end, that seems out of place, but I'm no RST expert

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 14]
```diff
@@ -62,29 +62,101 @@ In this section we outline steps needed for unit testing in Arrow.
...
+
+      In general any change to source code needs to be accompanied by unit tests. All tests are expected to pass before a pull request is merged.
+
+      * Add functionality -> add unit tests
+      * Modify functionality -> update unit tests
+      * Solve a bug -> add unit test before solving it, which helps prove the bug and its fix
+      * Performance improvements should be reflected in benchmarks (which are also tests)
+      * An exception could be refactoring functionality that is fully covered by unit tests
+
+      If the new functionality is a user-facing or API change, you will almost certainly need to change tests — if no tests need to be changed it might mean the tests aren't right! If the new functionality is a refactor and no APIs are changing, there might not need to be test changes.
```
Comment (jonkeane) [inline docs/source/developers/guide/step_by_step/testing.rst:142]:
```suggestion
      A good rule of thumb is: If the new functionality is a user-facing or API change, you will almost certainly need to change tests — if no tests need to be changed it might mean the tests aren't right! If the new functionality is a refactor and no APIs are changing, there might not need to be test changes.
```

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 15]
```diff
@@ -62,29 +62,101 @@ In this section we outline steps needed for unit testing in Arrow.
...
+      * ``skip_`` - skips a unit test - think of them as acceptable fails. Situations in which we might want to skip unit tests:
+
+        - ``skip_if_r_version()`` - this is a specific ``arrow`` skip. For example, we use this to skip a unit test when the R version is 3.5.0 and below (``skip_if_r_version(“3.5.0”)``). You will likely see it used when the functionality we are testing depends on features introduced after version 3.5.0 of R (such as the alternative representation of vectors, Altrep, introduced in R 3.5.0, but with significant additions in subsequent releases). As part of our CI workflow we test against different versions of R and this is where this feature comes in.
+        - ``skip_if_not_available()`` - another specific {arrow} skip. Arrow (libarrow) has a series of additional features that can be switched on or off (but this needs to happen at build time). If a unit test depends on such a feature and this feature is not available (i.e. was not selected when libarrow was built) the test is skipped, as opposed to having a failed test.
```
Comment (jonkeane) [inline docs/source/developers/guide/step_by_step/testing.rst:153]:
```suggestion
        - ``skip_if_not_available()`` - another specific {arrow} skip. Arrow (libarrow) has a number of optional features that can be switched on or off (which happens at build time). If a unit test depends on such a feature and this feature is not available (i.e. was not selected when libarrow was built) the test is skipped, as opposed to having a failed test.
```

File: docs/source/developers/guide/step_by_step/testing.rst [hunk 16]
```diff
@@ -62,29 +62,101 @@ In this section we outline steps needed for unit testing in Arrow.
...
+      * An exception could be refactoring functionality that is fully covered by unit tests
+
+      If the new functionality is a user-facing or API change, you will almost certainly need to change tests — if no tests need to be changed it might mean the tests aren't right! If the new functionality is a refactor and no APIs are changing, there might not need to be test changes.
+
+      **Testing helpers**
+
+      To complement the ``testthat`` functionality, the ``arrow`` R package has defined a series of specific utility functions (called helpers), such as:
+
+      * expectations - these start with ``expect_`` and are used to compare objects
+            - for example, ``expect_altrep_roundtrip()`` compares the result of a function ``fn`` run on a vector ``x`` with the result of the same function run on the altrep version of ``x``. More generally, ``expect_…_roundtrip()`` functions take an input, convert it to some other format (e.g. arrow) and then convert it back, confirming that the values are the same.
```
Comment (jonkeane) [inline docs/source/developers/guide/step_by_step/testing.rst:149]: I wonder if we should also provide an example here that shows a call? The two function arguments feel a bit out of place / floating out there: either we could drop them, or we should show them in use in `expect_altrep_roundtrip()`

**Review by jonkeane — CHANGES_REQUESTED** [review]
This is looking great. A few minor tweaks, and one suggestion

File: docs/source/developers/guide/step_by_step/testing.rst (hunk 13 above)
Comment (dragosmg) [reply docs/source/developers/guide/step_by_step/testing.rst:130]: I think the `_` is needed after `<>`

File: docs/source/developers/guide/step_by_step/testing.rst (hunk 15 above)
Comment (dragosmg) [reply docs/source/developers/guide/step_by_step/testing.rst:153]: Done

File: docs/source/developers/guide/step_by_step/testing.rst (hunk 16 above)
Comment (dragosmg) [reply docs/source/developers/guide/step_by_step/testing.rst:149]: Done. I changed the order of the sentences and went with general description of `expect_..._roundtrip()` + example

**Review by jonkeane — APPROVED** [review]
This looks good to me. Thank you!
